src/fides/api/ops/models/datasetconfig.py

```diff
@@ -40,12 +41,84 @@ class DatasetConfig(Base):
     dataset = Column(
         MutableDict.as_mutable(JSONB), index=False, unique=False, nullable=False
     )
+    ctl_dataset_id = Column(
+        String, ForeignKey(CtlDataset.id), index=True, nullable=False
+    )
 
     connection_config = relationship(
         ConnectionConfig,
         backref="datasets",
     )
 
+    ctl_dataset = relationship(
+        CtlDataset,
+        backref="dataset_configs",
+    )
+
+    @classmethod
+    def upsert_with_ctl_dataset(
+        cls, db: Session, *, data: Dict[str, Any]
+    ) -> "DatasetConfig":
+        """
+        Create or update the DatasetConfig AND the corresponding CTL Dataset
+
+        If the DatasetConfig exists with the supplied FidesKey, update the linked CtlDataset with the dataset contents.
+        If the DatasetConfig *does not exist*, upsert a CtlDataset on fides_key, and then link to the DatasetConfig on creation.
+
+        """
+
+        def upsert_ctl_dataset(ctl_dataset_obj: Optional[CtlDataset]) -> CtlDataset:
+            """
+            If ctl_dataset_obj specified, update that resource directly, otherwise
+            create a new resource.
+            """
+            ctl_dataset_data = data.copy()
+            validated_data = Dataset(**ctl_dataset_data.get("dataset", {}))
+
+            if ctl_dataset_obj:
+                # It's possible this updates the ctl_dataset.fides_key and this causes a conflict
+                # with another ctl_dataset, if we fetched the datasetconfig.ctl_dataset.
+                for key, val in ctl_dataset_data.get("dataset", {}).items():
+                    setattr(
+                        ctl_dataset_obj, key, val
+                    )  # Just update the existing ctl_dataset with the new values
+            else:
+                ctl_dataset_obj = CtlDataset(
+                    **validated_data.dict()
+                )  # Validate the values if creating a new CtlDataset
+
+            db.add(ctl_dataset_obj)
+            db.commit()
+            db.refresh(ctl_dataset_obj)
+            return ctl_dataset_obj
+
+        dataset = DatasetConfig.filter(
+            db=db,
+            conditions=(
+                (DatasetConfig.connection_config_id == data["connection_config_id"])
+                & (DatasetConfig.fides_key == data["fides_key"])
+            ),
+        ).first()
+
+        if dataset:
+            upsert_ctl_dataset(
+                dataset.ctl_dataset
+            )  # Update existing ctl_dataset first.
```

Comment:
Here, I know the specific ctl_dataset_id because I got it off the existing DatasetConfig. However, if there is no dataset config, I'm looking up a ctl dataset by fides key.

So there's a little extra code here, sometimes I want to update the CTLDataset by id, others by fides_key.